src/cljam/algo/pileup.clj

```diff
@@ -142,6 +145,155 @@
          (partial filterv)
          (update p 1))))
 
+(defn- merge-corrected-quals
+  "Merge corrected quals with the uncorrected part."
+  [^SAMAlignment aln corrected-quals]
+  (let [quals (:quals-at-ref aln)
+        pnext (.pnext aln)
+        start (.pos aln)
+        len   (count corrected-quals)]
+    (if (<= start pnext)
+      (into (into (vec (take (- pnext start) quals)) corrected-quals)
+            (drop (+ (- pnext start) len) quals))
+      (into corrected-quals (drop len quals)))))
+
+(defn- correct-pair-quals
+  "Correct quals of a pair. Returns a map with corrected quals."
+  [^SAMAlignment a1 ^SAMAlignment a2]
+  (if-not (and (pos? (.pnext a1)) (<= (.pnext a1) (.end a1)))
+    {}
+    (let [tlen1 (.tlen a1)
+          tlen2 (.tlen a2)
+          quals1 (:quals-at-ref a1)
+          quals2 (:quals-at-ref a2)
+          seqs1  (:seqs-at-ref a1)
+          seqs2  (:seqs-at-ref a2)
+          new-quals (for [pos (range (.pnext a1)
+                                     (inc (min (.end a1) (.end a2))))]
+                      (let [relative-pos1 (- pos (.pos a1))
+                            relative-pos2 (- pos (.pos a2))
+                            q1 (quals1 relative-pos1)
+                            q2 (quals2 relative-pos2)
+                            [b1 _] (seqs1 relative-pos1)
```

Comment:
@athos Could you show how to reproduce this issue?

I tried to make `new-quals` for only the overlapped range.
But it seems something wrong for my lack of consideration.

Reply:
Sorry for my late reply!

The link below shows two typical example cases that caused the error.

- https://gist.github.com/athos/8abbbd76ac250118f25d5d7073a9d8e7

Those pair reads are just artificial data, but made by manipulating some actual samples (I'm not sure whether the `:tlen` of the second pair is really correct since it looks weird, although it doesn't have to do with this issue).

If one read of a pair contains an insertion (or a soft clip), the length of `:quals-at-ref` and  `:seqs-at-ref` of the read will be shorter by the number of the inserted (or clipped) bases, compared to the other read.

But anyway, your latest changes seem to fix errors caused by my example cases 😃 One thing I would say is that I recommend using `.pos` both for `a1` and `a2` consistently (i.e. `(max (.pnext a1) (.pos a2))` should be `(max (.pos a1) (.pos a2))`) since the `.pnext` value is somewhat unreliable in some cases, as you said. Otherwise, `(- pos (.pos a1))` could point to a wrong index.

Reply:
Thanks for the suggestion!
I've fixed that.